File: web/common/src/main/java/org/artifactory/common/wicket/component/panel/feedback/FeedbackDistributer.js
/*
 * Artifactory is a binaries repository manager.
 * Copyright (C) 2012 JFrog Ltd.
 *
 * Artifactory is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Artifactory is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Artifactory.  If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * Distributes feedback messages to FeedbackMessagesPanel.
 *
 * @author Yoav Aharoni
 */
if (!window.FeedbackDistributer) {
    FeedbackDistributer = {
        defaultPanel:null,
        panels:new Array(),

        init:function (defaultPanel) {
            FeedbackDistributer.defaultPanel = dojo.byId(defaultPanel);
            FeedbackDistributer.clearMessages();
        },

        clearMessages:function () {
            if (FeedbackDistributer.panels.length == 0) {
                return;
            }
            dojo.forEach(FeedbackDistributer.panels, function (panel) {
                try {
                    while (panel.firstChild) {
                        var ondestroy = panel.getAttribute('ondestroy');
                        if (ondestroy) {
                            eval(ondestroy);
                        }
                        panel.removeChild(panel.firstChild);
                    }
                } catch (e) {
                }
            });

            FeedbackDistributer.panels = new Array();
        },

        showMessages:function () {
            dojo.forEach(FeedbackDistributer.panels, function (panel) {
                if (panel.getElementsByTagName('li').length) {
                    // trigger panel.onShow() event handler
                    var onshow = panel.getAttribute('onshow');
                    if (onshow) {
                        eval(onshow);
                    }
                }
            });
        },

        addMessage:function (reporterId, level, message) {
            // get reporter and feedback panel
            var reporter = dojo.byId(reporterId);
            var panel = FeedbackDistributer.getFeedbackPanelFor(reporter);

            // setup panel if needed
            if (!panel.hasChildNodes()) {
                var ul = document.createElement('ul');
                ul.className = 'feedback';
                panel.appendChild(ul);
                FeedbackDistributer.panels.push(panel);

                // trigger panel.onClear() event handler
                var onclear = panel.getAttribute('onclear');
                if (onclear) {
                    eval(onclear);
                }
            }

            // add message to panel
            var li = document.createElement('li');
            li.className = 'feedbackPanel' + level;
            li.innerHTML = '<span>' + message + '</span>';
            li.reporter = reporter;
            li.level = level;
            panel.firstChild.appendChild(li);
        },

        getFeedbackPanelFor:function (reporter) {
            var node = reporter;

            // find closest FeedbackPanel
            while (node && node != document.body) {
                var feedbackId = node.getAttribute('feedbackId');
                if (feedbackId) {
                    return dojo.byId(feedbackId);
                }

                node = node.parentNode;
            }

            // add defaultPanel if needed
            if (!FeedbackDistributer.defaultPanel) {
                var dp = document.createElement('div');
                document.body.appendChild(dp);
                FeedbackDistributer.defaultPanel = dp;
            }

            // return defaultPanel
            return FeedbackDistributer.defaultPanel;
        }
    };
}

File: web/common/src/main/java/org/artifactory/common/wicket/component/panel/feedback/FeedbackDistributer.java
/*
 * Artifactory is a binaries repository manager.
 * Copyright (C) 2012 JFrog Ltd.
 *
 * Artifactory is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Artifactory is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Artifactory.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.artifactory.common.wicket.component.panel.feedback;

import org.apache.wicket.Component;
import org.apache.wicket.feedback.FeedbackMessage;
import org.apache.wicket.feedback.FeedbackMessagesModel;
import org.apache.wicket.feedback.IFeedback;
import org.apache.wicket.markup.html.basic.Label;
import org.apache.wicket.markup.html.panel.Panel;
import org.apache.wicket.model.AbstractReadOnlyModel;
import org.apache.wicket.util.string.Strings;
import org.artifactory.common.wicket.contributor.ResourcePackage;

import java.io.Serializable;
import java.util.Collections;
import java.util.List;

/**
 * @author Yoav Aharoni
 */
@SuppressWarnings({"BooleanMethodNameMustStartWithQuestion"})
public class FeedbackDistributer extends Panel implements IFeedback {
    private FeedbackMessagesPanel defaultFeedbackPanel;

    public FeedbackDistributer(String id) {
        super(id);

        setDefaultModel(newFeedbackMessagesModel());
        setOutputMarkupId(true);

        add(ResourcePackage.forJavaScript(FeedbackDistributer.class));

        Label script = new Label("script", new ScriptModel());
        script.setEscapeModelStrings(false);
        script.setVersioned(false);
        add(script);
    }

    public FeedbackMessagesPanel getDefaultFeedbackPanel() {
        return defaultFeedbackPanel;
    }

    public void setDefaultFeedbackPanel(FeedbackMessagesPanel defaultFeedbackPanel) {
        this.defaultFeedbackPanel = defaultFeedbackPanel;
    }

    /**
     * @see Component#isVersioned()
     */
    @SuppressWarnings({"RefusedBequest"})
    @Override
    public boolean isVersioned() {
        return false;
    }

    public final FeedbackMessagesModel getFeedbackMessagesModel() {
        return (FeedbackMessagesModel) getDefaultModel();
    }

    public void updateFeedback() {
        // Force model to load
        getDefaultModelObject();
    }

    public boolean anyErrorMessage() {
        return anyMessage(FeedbackMessage.ERROR);
    }

    public boolean anyMessage() {
        return anyMessage(FeedbackMessage.UNDEFINED);
    }

    public boolean anyMessage(int level) {
        List<FeedbackMessage> msgs = getCurrentMessages();

        for (FeedbackMessage msg : msgs) {
            if (msg.isLevel(level)) {
                return true;
            }
        }

        return false;
    }

    @SuppressWarnings({"unchecked"})
    protected final List<FeedbackMessage> getCurrentMessages() {
        List<FeedbackMessage> messages = (List<FeedbackMessage>) getDefaultModelObject();
        return Collections.unmodifiableList(messages);
    }

    protected FeedbackMessagesModel newFeedbackMessagesModel() {
        return new FeedbackMessagesModel(this);
    }

    private class ScriptModel extends AbstractReadOnlyModel {
        @Override
        public Object getObject() {
            StringBuffer initScript = new StringBuffer();

            // init
            String defaultFeedbackId = defaultFeedbackPanel == null
                    ? "" : defaultFeedbackPanel.getMarkupId();

            initScript
                    .append("(function(fd) {\n")
                    .append("fd.init('")
                    .append(defaultFeedbackId)
                    .append("');\n");

            // add messages
            for (FeedbackMessage feedbackMessage : getCurrentMessages()) {
                feedbackMessage.markRendered();
                Serializable message = feedbackMessage.getMessage();

                String messageString = (message instanceof UnescapedFeedbackMessage) ?
                        message.toString() : Strings.escapeMarkup(message.toString(), false, false).toString();
                initScript
                        .append("fd.addMessage('")
                        .append(getMarkupIdFor(feedbackMessage.getReporter()))
                        .append("', '")
                        .append(feedbackMessage.getLevelAsString())
                        .append("', ")
                        .append(asJsStringParam(messageString))
                        .append(");\n");
            }

            initScript.append("fd.showMessages();\n");
            initScript.append("})(FeedbackDistributer);\n");
            return initScript;
        }

        private String getMarkupIdFor(Component component) {
            Component current = component;

            // find first parent with markup id
            while (current != null) {
                if (current.getOutputMarkupId() &&
                        (!current.getRenderBodyOnly() || current.getOutputMarkupPlaceholderTag())) {
                    return current.getMarkupId();
                }

                current = current.getParent();
            }

            return "";
        }
    }

    public static String asJsStringParam(Object param) {
        return '\'' + param.toString()
                .replaceAll("\\\\", "\\\\\\\\")
                .replaceAll("'", "\\\\'")
                .replaceAll("\n", "<br/>")
                .replaceAll("\t", "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;")
                .replaceAll("\r", "") + '\'';
    }
}
